<launch>
  <!-- Start the multi-agent navigation stack -->
  <param name="use_sim_time" value="true"/>
  
  <!-- Launch Gazebo with the specified world -->
  <arg name="robot_type" default="jackal"/>
  <arg name="use_sim_time" default="true" />
  <arg name="headless" default="true" />
  <arg name="verbose" default="true" />
  <arg name="paused" default="false"/>
  <arg name="debug" default="false" />
  <arg name="gui" default="true" />
  <arg name="world_name" default="$(find multi_agent_formation)/worlds/square_obstacle_more.world"/>
  <!-- <arg name="world_name" default="$(find jackal_helper)/worlds/jackal_race.world" /> -->

  <include file="$(find gazebo_ros)/launch/empty_world.launch">
    <arg name="use_sim_time" value="$(arg use_sim_time)"/>
    <arg name="headless" value="$(arg headless)"/>
    <arg name="verbose" value="$(arg verbose)"/>
    <arg name="paused" value="$(arg paused)"/>
    <arg name="debug" value="$(arg debug)"/>
    <arg name="gui" value="$(arg gui)"/>
    <arg name="world_name" value="$(arg world_name)" />
  </include>

  <!-- Launch the training system -->
  <!-- <node pkg="multi_agent_formation" name="multi_agent_formation" type="leader_controller.py" output="screen"/> -->

  <!-- tf -->
  <!-- <node pkg="tf" type="static_transform_publisher" name="odom_1_map_broadcaster" args="0.0 0.0 0.0 0 0 0.0 map robot_1/odom 100" /> -->
  <!-- <node pkg="tf" type="static_transform_publisher" name="odom_2_map_broadcaster" args="0.0 0.0 0.0 0 0 0.0 map robot_2/odom 100" /> -->
  <!-- <node pkg="tf" type="static_transform_publisher" name="odom_3_map_broadcaster" args="0.0 0.0 0.0 0 0 0.0 map robot_3/odom 100" /> -->
  <!-- <node pkg="tf" type="static_transform_publisher" name="odom_4_map_broadcaster" args="0.0 0.0 0.0 0 0 0.0 map robot_4/odom 100" /> -->

  <!-- Load turtlebot -->
  <group ns="robot_1">
    <arg name="robot_namespace" value="robot_1" />
    <param name="tf_prefix" value="robot_1" />
    <include file="$(find multi_agent_formation)/launch/spawn_$(arg robot_type).launch">
        <arg name="robot_namespace" value="$(arg robot_namespace)" />
        <!-- <arg name="tf_prefix" value="$(arg tf_prefix)" /> -->
        <arg name="x" value="-1.0" />
        <arg name="y" value="1.0" />
        <arg name="yaw" value="0.0" />
        <arg name="joystick" default="false" />
    </include>
    
    <!-- The configuration files are loaded from multi_agent_formation -->
    <node pkg="move_base" type="move_base" respawn="false" name="move_base" output="screen">

      <rosparam file="$(find multi_agent_formation)/configs/costmap_common_params.yaml" command="load" ns="global_costmap" />
      <rosparam file="$(find multi_agent_formation)/configs/costmap_common_params.yaml" command="load" ns="local_costmap" />

      <rosparam file="$(find multi_agent_formation)/configs/odom_nav_params/global_costmap_params.yaml" command="load" />
      <rosparam file="$(find multi_agent_formation)/configs/odom_nav_params/local_costmap_params.yaml" command="load" />

      <rosparam file="$(find multi_agent_formation)/configs/teb_local_planner_params.yaml" command="load" />
      <rosparam file="$(find multi_agent_formation)/configs/move_base_params.yaml" command="load" />
      <rosparam file="$(find multi_agent_formation)/configs/global_planner_params.yaml" command="load" />

      <param name="base_global_planner" value="global_planner/GlobalPlanner" />
      <param name="planner_frequency" value="1.0" />
      <param name="planner_patience" value="5.0" />

      <param name="base_local_planner" value="teb_local_planner/TebLocalPlannerROS" />
      <param name="controller_frequency" value="5.0" />
      <param name="controller_patience" value="15.0" />
      
      <!-- <remap from="move_base_simple/goal" to="robot_1/move_base_simple/goal" /> -->

    </node>    
    <!-- <remap from="cmd_vel" to="robot_1/cmd_vel" /> -->
  </group>

  <!-- <group ns="robot_2">
    <include file="$(find multi_agent_formation)/launch/spawn_$(arg robot_type).launch">
        <arg name="robot_namespace" value="/robot_2" />
        <arg name="tf_prefix" value="robot_2" />
        <arg name="x" value="1.0" />
        <arg name="y" value="-1.0" />
        <arg name="yaw" value="0.0" />
        <arg name="joystick" default="false" />
    </include>
  </group>

  <group ns="robot_3">
    <include file="$(find multi_agent_formation)/launch/spawn_$(arg robot_type).launch">
        <arg name="robot_namespace" value="/robot_3" />
        <arg name="tf_prefix" value="robot_3" />
        <arg name="x" value="-1.0" />
        <arg name="y" value="-1.0" />
        <arg name="yaw" value="0.0" />
        <arg name="joystick" default="false" />
    </include>
  </group>

  <group ns="robot_4">
    <include file="$(find multi_agent_formation)/launch/spawn_$(arg robot_type).launch">
        <arg name="robot_namespace" value="/robot_4" />
        <arg name="tf_prefix" value="robot_4" />
        <arg name="x" value="1.0" />
        <arg name="y" value="1.0" />
        <arg name="yaw" value="0.0" />
        <arg name="joystick" default="false" />
    </include>
  </group> -->
  
  <node pkg="rviz" type="rviz" name="rviz" args="-d $(find teb_local_planner)/cfg/rviz_teb_sim.rviz" required="true" />

</launch>
